### ISSUE TITLE
Provided main is ignored and the default main fails if packageName is not included

### ISSUE THREAD

**Issue by justinbmeyer** [donejs/done-ssr-middleware#25]
@justinbmeyer commented on [Thu Jun 15 2017](https://github.com/donejs/donejs/issues/965)

It seems that `ssr` is failing if the `main` is like `"main": "index.stache!done-autorender",`.  

I think this is a bug b/c this works fine if you load StealJS in production as a __client-rendered__ app.  It only breaks with SSR.

---

@justinbmeyer commented on [Thu Jun 15 2017](https://github.com/donejs/donejs/issues/965#issuecomment-308908140)

This is breaking even if I call `ssr` with the right values:

```js
module.exports = ssr({
  config: __dirname + "/package.json!npm",
  main: "bitballs/index.stache!done-autorender",
  liveReload: true
});
```

---

@justinbmeyer commented on [Thu Jun 15 2017](https://github.com/donejs/donejs/issues/965#issuecomment-308909099)

Looks like `done-ssr-middleware` ignores the passed in configuration:

```js
// In production we need to pass in the main, otherwise it doesn't know what
// bundle to load from.
if(process.env.NODE_ENV === 'production') {
		system.main = (pkg.system && pkg.system.main) || (pkg.steal && pkg.steal.main) || pkg.main;
}
```

If `system.main` is present, this should probably assume the person knows right. Also, it might want to include the `package.name` here in the main if it's not already present.

**Comment by matthewp** [comment]
This issue seems to have changed a couple of times. Is the issue now that it is ignoring your passed in `main` value and using from the package.json instead?

**Comment by justinbmeyer** [comment]
@matthewp yeah, I guess there are two issues actually.  The easiest one to fix is that it's ignoring the passed in `main`.  The other issue is that it creates a main w/o the `package.name`.

**Comment by matthewp** [comment]
> The other issue is that it creates a main w/o the package.name.

Can you explain with an example?

**Comment by justinbmeyer** [comment]
Yes, [currently](https://github.com/donejs/bitballs/blob/222441dc2fcce02f323b0a23cdab2081c1316161/public/package.json#L11) bitball's master looks like this:

```
"main": "index.stache!done-autorender",
```

It doesn't have a `system.steal.main` as well.

Running ssr in production fails.

**Comment by matthewp** [comment]
Why would you expect that to work though? The bundle is presumably located at `dist/bundles/bitballs/index.js`, how would it know where to find that if you exclude the package name?

**Comment by justinbmeyer** [comment]
@matthewp The `.main` is in _package.json_ ... it seems like it should be able to read it's own `.name`

**Comment by matthewp** [comment]
Ah, I see. I'm not a huge fan of that kind of error correction personally. I agree that users shouldn't have to understand the complexity of the normalize algorithm, but that's something we have to fix in Steal.

**Comment by justinbmeyer** [comment]
I'm talking about this code: https://github.com/donejs/done-ssr-middleware/blob/master/lib/index.js#L23

It has the whole package.json.


IMO, this isn't error correctly.  I don't expect to have to put my package name in my `main`.  I think that is what we should generate with.

**Comment by matthewp** [comment]
From a DoneJS perspective I would find it confusing as a user if `index.stache!done-autorender` worked in SSR in production but didn't work in Steal or in steal-tools, or any of the other places where you might have to provide a module name.

**Comment by justinbmeyer** [comment]
I can see that, but from a node perspective, it's quite odd.

**Comment by matthewp** [comment]
Agree it's odd, typically in Node you would point to a specific file. If we could make that work I think it would be a better api.

**Comment by matthewp** [comment]
For now I'm going to fix the issue with `main` being ignored, can contain to discuss better api though.

**Comment by justinbmeyer** [comment]
if nothing else, we should give users a warning.  This took a long time to figure out because dev and production __client rendered__ both worked.  I think the inconsistency means there's a bug.

**Comment by matthewp** [comment]
I'd be surprised if using that `main` worked in client-rendered production. Without the package name it wouldn't know to look up the bundle in a subfolder.  I'd love to see an example of that working in client-rendered production.

**Comment by justinbmeyer** [comment]
I think we must be talking about different things. This type of main is what's expected for stealjs. https://stealjs.com/docs/StealJS.guides.progressive_loading.html#update-packagejson-with-the-right-main

Why would DoneJS need to be different?

**Comment by justinbmeyer** [comment]
To be clear I'm talking about what package.json's main is. NOT the main in the script tag.

**Comment by matthewp** [comment]
#28 is submitted to fix the user-provided main issue.

I think the "the default main fails if packageName is not included" is not fixable without scrubbing the filesystem.

The reason is that both of these are valid ways to build:

```js
stealTools.build({
  config: __dirname + "/path/to/package.json!npm",
  main: "index.stache!done-autorender"
});
```

Which will write out a bundle to `dist/bundles/index.js`

```js
stealTools.build({
  config: __dirname + "/path/to/package.json!npm",
  main: "project/index.stache!done-autorender"
});
```

Which will write out a bundle to `dist/bundles/project/index.js`.

So the proposed change will break the first case. The only way I can think to fix this is to scrub the filesystem and try to determine where the main bundle is actually located.

cc @justinbmeyer

**Comment by matthewp** [comment]
Moving the discussion about how to find the main bundle to https://github.com/donejs/done-ssr-middleware/issues/29

**Comment by matthewp** [comment]
Released as 1.0.2 https://github.com/donejs/done-ssr-middleware/releases/tag/v1.0.2
